### DISCUSSION TITLE
nonempty vs not null + DATE type

### DISCUSSION THREAD

**Discussion by jkdoshi** [pocketbase/pocketbase#1256] [Q&A] [answered]
Recently we seem to have shifted from "not null" to "nonempty" fields. But it seems to have really weird effects on the field values. Now when I mark a field nonempty, I'm not able to store 0 as a value. Does that seem right?

Also, when I go into sqlite3 CLI and examine the underlying table, the date fields are typed as TEXT instead of DATE or DATETIME. That also seems weird. And it has real consequences for my application, because now when I look at the data using a BI tool, the tool refuses to aggregate the date by week, month, quarter, year, etc because the underlying column is TEXT instead of DATE.

**Comment by ganigeorgiev** [comment, marked as answer]
> Now when I mark a field nonempty, I'm not able to store 0 as a value. Does that seem right?

Yes, that is expected. Since PocketBase v0.3+ the fields (_with exception to the `json` type_) use zero-defaults. For a full list with the supported values you can check https://pocketbase.io/docs/collections/#overview.

There should be a tooltip alongside the label of "Nonempty" that explains what field values the constraint will require.

---

> Also, when I go into sqlite3 CLI and examine the underlying table, the date fields are typed as TEXT instead of DATE or DATETIME.

SQLite3 doesn't have date/datetime datatypes - https://www.sqlite.org/datatype3.html#date_and_time_datatype.

**Reply by jkdoshi** [reply]
> There should be a tooltip alongside the label of "Nonempty" that explains what field values the constraint will require.

Okay. But then how to have numeric fields whose value is required (not null) but allows 0 as a valid value?

> SQLite3 doesn't have date/datetime datatypes - https://www.sqlite.org/datatype3.html#date_and_time_datatype.

Hmm ... that's strange because I range the following on my field ...

```
ALTER TABLE txns DROP COLUMN date; -- sqlite does not allow changing a column type, you have to drop and add
ALTER TABLE txns ADD COLUMN date DATE; -- sqlite allows this without errors
```
And now my BI tool _actually_ treats my `date` field as a temporal field.

And also, examining the schema shows that the column is really a DATE now.
```
.schema txns
--- output below ...
CREATE TABLE IF NOT EXISTS "txns" (
  ...
  `created` TEXT DEFAULT "" NOT NULL,
  `date` DATE,
  `description` TEXT DEFAULT '',
  ...
);
```

**Reply by ganigeorgiev** [reply]
> But then how to have numeric fields whose value is required (not null) but allows 0 as a valid value?

PocketBase doesn't store `null` for the field value (with exception of the `json` type field).
Even if the user submit a `null` it will be casted to 0. If you want to accept `0` as value don't mark the field as "Nonempty".

> Hmm ... that's strange because I range the following on my field ...

It doesn't really matter what you define as table column type in the schema. SQLite even allow storing mixed values in the same column (unless strict is set).

What BI tool are you using and how do you load the sqlite data into it?


> And also, examining the schema shows that the column is really a DATE now.

The schema definition is just a text string. You can type whatever you want in it, even the spacing will be preserved.

**Reply by ganigeorgiev** [reply]
Maybe the BI tool is reading the schema definition and tries to detect the column type from it? I'm not sure but for SQLite it shouldn't matter. If you want to operate with date/datetime you have to use the [sqlite datetime functions](https://www.sqlite.org/lang_datefunc.html).

**Reply by jkdoshi** [reply]
First of all, thank you for the real-time responses. I feel guilty that I'm distracting you from "real work"

> PocketBase doesn't store null for the field value (with exception of the json type field).
> Even if the user submit a null it will be casted to 0. If you want to accept 0 as value don't mark the field as "Nonempty".

Is that a pocketbase thing or sqlite thing? I tested it, and sqlite does very much differentiate NULL from 0 for columns of type "REAL NOT NULL" (allows 0 but not null).

> What BI tool are you using and how do you load the sqlite data into it?

I'm using [Metabase](https://www.metabase.com/).

> Maybe the BI tool is reading the schema definition and tries to detect the column type from it?

Quite likely.

But then if putting TEXT or DATE in the column type makes no difference to sqlite then is it not doing any validation that the data type is correct?

All this leads me to a different question -- to what extent can I mess with the tables that pocketbase generates? And what the correct way of doing it? I'd like to do it in a way that plays well with sqlite as well as pocketbase.

**Reply by ganigeorgiev** [reply]
> Is that a pocketbase thing or sqlite thing? I tested it, and sqlite does very much differentiate NULL from 0 for columns of type "REAL NOT NULL" (allows 0 but not null).

It is a PocketBase thing. Not having `null` removes the ambiguity when creating records with missing body fields and makes it a lot easier to work with the response in "non-nullable by default" languages like Dart.

---

>But then if putting TEXT or DATE in the column type makes no difference to sqlite then is it not doing any validation that the data type is correct?

Yes, that's correct at least for non-strict defined tables. SQLite has dynamic types or "type affinity":

> The type affinity of a column is the recommended type for data stored in that column. The important idea here is that the type is recommended, not required. **Any column can still store any type of data.**
>
> If the declared type contains the string "INT" then it is assigned INTEGER affinity.
>
> If the declared type of the column contains any of the strings "CHAR", "CLOB", or "TEXT" then that column has TEXT affinity. Notice that the type VARCHAR contains the string "CHAR" and is thus assigned TEXT affinity.
>
> If the declared type for a column contains the string "BLOB" or if no type is specified then the column has affinity BLOB.
>
> If the declared type for a column contains any of the strings "REAL", "FLOA", or "DOUB" then the column has REAL affinity.
> 
> Otherwise, the affinity is NUMERIC.
>
> Note that the order of the rules for determining column affinity is important. A column whose declared type is "CHARINT" will match both rules 1 and 2 but the first rule takes precedence and so the column affinity will be INTEGER.

---

> to what extent can I mess with the tables that pocketbase generates? And what the correct way of doing it?

You shouldn't manually change the table schema of a PocketBase created table, since this could conflict with future migration scripts.

---

There are several issues in metabase related to the SQLite datetime handling, for example you could check https://discourse.metabase.com/t/handling-date-timestamps-in-sqllite/7188/4.
Technically we can change the date column type to `DATETIME` but its affinity will be incorrect since it will be `NUMERIC` according to the rules above and not ISO86001 strings.

**Comment by jkdoshi** [comment]
Thank you so much for taking time to answer in such detail. I've marked it as answered.
